cmd/logger/audit.go

```diff
@@ -27,23 +30,43 @@ import (
 // ResponseWriter - is a wrapper to trap the http response status code.
 type ResponseWriter struct {
 	http.ResponseWriter
-	statusCode      int
+	StatusCode      int
 	startTime       time.Time
 	timeToFirstByte time.Duration
+	// number of bytes written
+	bytesWritten int
+	// Response body should be logged
+	LogBody bool
+	// Internal recording buffer
+	headers bytes.Buffer
+	body    bytes.Buffer
+	// Indicate if headers are written in the log
+	headersLogged bool
 }
 
 // NewResponseWriter - returns a wrapped response writer to trap
 // http status codes for auditiing purposes.
 func NewResponseWriter(w http.ResponseWriter) *ResponseWriter {
 	return &ResponseWriter{
 		ResponseWriter: w,
-		statusCode:     http.StatusOK,
+		StatusCode:     http.StatusOK,
 		startTime:      time.Now().UTC(),
 	}
 }
 
 func (lrw *ResponseWriter) Write(p []byte) (int, error) {
 	n, err := lrw.ResponseWriter.Write(p)
+	lrw.bytesWritten += n
+	if !lrw.headersLogged {
+		// We assume the response code to be '200 OK' when WriteHeader() is not called,
+		// that way following Golang HTTP response behavior.
+		lrw.writeHeaders(&lrw.headers, http.StatusOK, lrw.Header())
+		lrw.headersLogged = true
+	}
+	if lrw.StatusCode >= http.StatusBadRequest || lrw.LogBody {
+		// Always logging error responses.
+		lrw.body.Write(p)
+	}
```

Comment:
timeToFirstByte might have slight timing mistakes here for example time taken to fill the bytes.Buffer might get added for large headers and bodies, which may not capture the right timeToFirstByte perhaps the logic

```
if lrw.timeToFirstByte == 0 {
		lrw.timeToFirstByte = time.Now().UTC().Sub(lrw.startTime)
	}
```

should be done right at L59